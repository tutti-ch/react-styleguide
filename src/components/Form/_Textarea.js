import React, { Component } from "react";
import PropTypes from "prop-types";
import { filterProps } from "../../helpers/functions";
import WithWrapper from "./_WithWrapper";

export class Textarea extends Component {
  static defaultProps = {
    value: ""
  };

  static propTypes = {
    /**
     * The initial value.
     */
    value: PropTypes.string,

    /**
     * The input name.
     */
    name: PropTypes.string
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };

    this.notify = this.notify.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  /**
   * Sync the value.
   */
  componentDidUpdate(oldProps) {
    if (oldProps.value !== this.props.value) {
      this.setState({ value: this.props.value });
    }
  }

  /**
   * Handle the change event.
   */
  handleChange(event) {
    this.setState({ value: event.target.value }, () => {
      this.changed = true;
    });
  }

  /**
   * Notify the parent of the change.
   */
  notify() {
    if (this.changed === true && typeof this.props.onChange === "function") {
      const { name, value } = this.props;
      const newValue = this.state.value;
      this.props.onChange(newValue, {
        name,
        formValue: newValue,
        initialValue: value
      });
    }

    // Reset the value
    this.changed = false;
  }

  /**
   * Notify the parent when blur occurs.
   *
   * @param event
   */
  handleBlur(event) {
    this.notify();

    // istanbul ignore else
    if (typeof this.props.onBlur === "function") {
      this.props.onBlur(event);
    }
  }

  /**
   * Notify the parent when user presses enter.
   *
   * @param event
   */
  handleKeyUp(event) {
    if (event.key === "Enter") {
      this.notify();
    }

    if (typeof this.props.onKeyUp === "function") {
      this.props.onKeyUp(event);
    }
  }

  render() {
    const { value } = this.state;
    const { name } = this.props;
    const props = filterProps(Textarea.propTypes, this.props);

    return (
      <textarea
        {...props}
        name={name}
        value={value}
        onBlur={this.handleBlur}
        onKeyUp={this.handleKeyUp}
        onChange={this.handleChange}
      />
    );
  }
}

export default WithWrapper(Textarea);
