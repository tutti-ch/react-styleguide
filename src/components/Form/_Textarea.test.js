/* global describe, test, expect, jest */
import React from "react";
import { Textarea } from "./_Textarea";
import { shallow, mount } from "enzyme";

describe("(Component) Textarea", () => {
  test("[componentDidUpdate] should sync the state with props", () => {
    const comp = shallow(<Textarea />);
    const inst = comp.instance();
    inst.setState = jest.fn();
    comp.setProps({ value: "Frocello" });
    expect(inst.setState).toMatchSnapshot();
  });

  test("[handleChange] should set the value and call the onChange from parent", () => {
    const comp = shallow(<Textarea />);
    const inst = comp.instance();
    inst.handleChange({ target: { value: "Moreno" } });
    expect(comp.state("value")).toBe("Moreno");
  });

  test("[handleBlur] should notify the parent and call attached onBlur", () => {
    const onBlur = jest.fn();
    const onChange = jest.fn();
    const comp = shallow(<Textarea onBlur={onBlur} onChange={onChange} />);
    const inst = comp.instance();
    const event = {};
    inst.handleBlur(event);
    expect(onBlur).toHaveBeenCalledWith(event);
    expect(onChange).not.toHaveBeenCalled(); // Because nothing changed.
  });

  test("[handleKeyUp] should notify the parent and call attached onKeyUp", () => {
    const onKeyUp = jest.fn();
    const onChange = jest.fn();
    const comp = shallow(<Textarea onKeyUp={onKeyUp} onChange={onChange} />);
    const inst = comp.instance();
    inst.handleChange({ target: { value: "abc" } });
    inst.handleKeyUp({ key: "Enter" });
    expect(onKeyUp).toHaveBeenCalledWith({ key: "Enter" });
    expect(onChange).toHaveBeenCalled();

    comp.setProps({ onKeyUp: undefined });
    inst.handleChange({ target: { value: "def" } });
    inst.handleKeyUp({ key: "ArrowDown" });
    expect(onKeyUp).toHaveBeenCalledTimes(1); // Because we have set to undefined
    expect(onChange).toHaveBeenCalledTimes(1); // Because they key up was not an enter ðŸ˜€
  });

  test("snapshot", () => {
    const comp = mount(<Textarea value="Life is good" type="text" />);
    expect(comp).toMatchSnapshot();
  });
});
