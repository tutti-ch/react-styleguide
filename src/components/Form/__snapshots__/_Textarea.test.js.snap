// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Textarea snapshot 1`] = `
<Textarea
  type="text"
  value="Life is good"
>
  <textarea
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="Life is good"
  />
</Textarea>
`;
